src/main/scala/is/hail/distributedmatrix/HailBlockMatrix.scala

```diff
@@ -0,0 +1,587 @@
+  }
+
+  def add(other: M): M =
+    blockMap2(other, _ + _)
+
+  def subtract(other: M): M =
+    blockMap2(other, _ - _)
+
+  def pointwiseMultiply(other: M): M =
+    blockMap2(other, _ :* _)
+
+  def pointwiseDivide(other: M): M =
+    blockMap2(other, _ :/ _)
+
+  def multiply(other: M): M =
+    new HailBlockMatrix(new HailBlockMatrixMultiplyRDD(this, other), blockSize, rows, other.cols)
+
+  def multiply(r: BDM[Double]): M = {
+    require(this.cols == r.rows,
+      s"incompatible matrix dimensions: ${ this.rows }x${ this.cols } and ${ r.rows }x${ r.cols }")
+    multiply(HailBlockMatrix.from(this.blocks.sparkContext, r, this.blockSize))
+  }
+
+  def scalarAdd(i: Double): M =
+    this.blockMap(_ + i)
+
+  def scalarSubtract(i: Double): M =
+    this.blockMap(_ - i)
+
+  def scalarMultiply(i: Double): M =
+    this.blockMap(_ :* i)
+
+  def scalarDivide(i: Double): M =
+    this.blockMap(_ / i)
+
+  def vectorAddToEveryColumn(v: Array[Double]): M = {
+    require(v.length == this.rows, s"vector length, ${ v.length }, must equal number of matrix rows ${ this.rows }; v: $v, m: $this")
+    val vbc = this.blocks.sparkContext.broadcast(v)
+    this.mapWithIndex((i, j, x) => x + vbc.value(i.toInt))
+  }
+
+  def vectorPointwiseMultiplyEveryColumn(v: Array[Double]): M = {
+    require(v.length == this.rows, s"vector length, ${ v.length }, must equal number of matrix rows ${ this.rows }; v: $v, m: $this")
+    val vbc = this.blocks.sparkContext.broadcast(v)
+    this.mapWithIndex((i, j, x) => x * vbc.value(i.toInt))
+  }
+
+  def vectorPointwiseMultiplyEveryRow(v: Array[Double]): M = {
+    require(v.length == this.cols, s"vector length, ${ v.length }, must equal number of matrix columns ${ this.cols }; v: $v, m: $this")
+    val vbc = this.blocks.sparkContext.broadcast(v)
+    this.mapWithIndex((i, j, x) => x * vbc.value(j.toInt))
+  }
+
+  def cache(): this.type = {
+    blocks.cache()
+    this
+  }
+
+  def persist(storageLevel: StorageLevel): this.type = {
+    blocks.persist(storageLevel)
+    this
+  }
+
+  def toLocalMatrix(): BDM[Double] = {
+    require(this.rows < Int.MaxValue, "The number of rows of this matrix should be less than " +
+      s"Int.MaxValue. Currently numRows: ${ this.rows }")
+    require(this.cols < Int.MaxValue, "The number of columns of this matrix should be less than " +
+      s"Int.MaxValue. Currently numCols: ${ this.cols }")
+    require(this.rows * this.cols < Int.MaxValue, "The length of the values array must be " +
+      s"less than Int.MaxValue. Currently rows * cols: ${ this.rows * this.cols }")
+    val rows = this.rows.toInt
+    val cols = this.cols.toInt
+    val localBlocks = blocks.collect()
+    val values = new Array[Double](rows * cols)
+    var bi = 0
+    while (bi < localBlocks.length) {
+      val ((blocki, blockj), m) = localBlocks(bi)
+      val ioffset = blocki * blockSize
+      val joffset = blockj * blockSize
+      m.foreachPair { case ((i, j), v) =>
```

Comment:
Breeze [claims unsafeValueAt is unnecessary](https://github.com/scalanlp/breeze/blob/master/math/src/main/scala/breeze/linalg/DenseMatrix.scala#L218). I'm inclined not to add complexity here since this operation will pale in comparison to the cost of a multiplication.